### PR TITLE
add CLI for Workflows management

### PR DESCRIPTION
**Pull request by perf2711**
This PR adds a CLI that allows to manage workflows objects:

```
morgue workflows connection [create | list | get | update | delete] <options>
morgue workflows integration [create | list | get | update | delete] <options>
morgue workflows alert [create | list | get | update | delete] <options>
```

Inspired heavily by existing alerts CLI.

Jira task: BT-2130

### CODE REVIEW

**Review by konraddysput — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1525,16 +1585,389 @@ The `--trigger` option has the form:
...
+
+#### List connections
+
+```
+morgue workflows connection list [options]
+```
+
+Options:
+
+- `--raw` - `boolean` - output raw JSON
```
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md:1561]: as it is a switch, `false`

File: README.md [hunk 2]
```diff
@@ -1525,16 +1585,389 @@ The `--trigger` option has the form:
...
+
+By using `--from-file`, you can load connection spec from file.
+Arguments provided by CLI will override the file spec, but they may not be required anymore.
+
+Options:
+
+- `--name` - `string` - connection name, required
+- `--plugin` - `string` plugin ID, required
+- `--options` - `object` - connection options specific to plugin ID, required
+- `--from-file` - `string` - load connection spec from file
```
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md:1599]: >By using `--from-file`, you can load connection spec from file.
>Arguments provided by CLI will override the file spec, but they may not be required anymore.

File: lib/workflows/connections.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    output(
+      connections.sort((c1, c2) => c1.name.localeCompare(c2.name)),
+      argv,
+      printConnection
+    );
+  }
+
+  async createConnection(argv) {
+    const init = loadInit(argv);
```
[the remark this answers is not in the document]
Comment (perf2711) [reply lib/workflows/connections.js:51]: The plugin ID may be specified in a file.

Also, really? You're not loading a 10GB JSON, it's a very small file with settings.

File: README.md [hunk 4]
```diff
@@ -1525,16 +1585,389 @@ The `--trigger` option has the form:
...
+morgue workflows connection create <options>
+```
+
+By using `--from-file`, you can load connection spec from file.
+Arguments provided by CLI will override the file spec, but they may not be required anymore.
+
+Options:
+
+- `--name` - `string` - connection name, required
+- `--plugin` - `string` plugin ID, required
```
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md:1597]: We do not. There is no spec also for the integration/connection options.

I can write something up for S3 Export, but I would not do this for rest of the plugins right now.

File: lib/workflows/models/createAlert.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            argv.condition || init.condition
+          ),
+          state: cliOptions.convertAtMostOne("state", argv.state || init.state),
+          filters: argv.filter
+            ? cliOptions
+                .convertMany("filter", argv.filter, true)
+                .map(parseFilter)
+                .map((filter) => ({ type: "attribute", ...filter }))
+            : cliOptions.convertMany("filter", init.filters, true),
+          threshold: cliOptions.convertAtMostOne(
```
[the remark this answers is not in the document]
Comment (perf2711) [reply lib/workflows/models/createAlert.js:44]: done

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (konraddysput) [reply README.md:1599]: its not obcious based on the options description

File: lib/workflows/connections.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (konraddysput) [reply lib/workflows/connections.js:51]: re: 10GB JSON 

it doesn't matter if you deal with 10GB JSON or a stream. it is always better to just validate everything you have in memory and then user resources that take time to load. file operation is one of them. It's just a good practice. 

I missed pluginId may be specified in the file

File: lib/workflows/integrations.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      this.project,
+      idOrName
+    );
+
+    output(integration, argv, printIntegration);
+  }
+}
+
+function printIntegration(integration) {
+  console.log(integration.id);
```
Comment (wca) [inline lib/workflows/integrations.js:123]: Could this be worth prefixing with a name so it's clear what the ID is for?

File: lib/workflows/client.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      method: "PUT",
+      path: `/universes/${universe}/projects/${project}/integrations/${idOrName}`,
+      body: integration,
+    });
+  }
+
+  deleteIntegration(universe, project, idOrName) {
+    return this.request({
+      method: "DELETE",
+      path: `/universes/${universe}/projects/${project}/integrations/${idOrName}`,
```
Comment (wca) [inline lib/workflows/client.js]: [nit] These URL paths could be DRY.

Comment (perf2711) [reply lib/workflows/client.js]: done

File: lib/workflows/integrations.js (hunk 6 above)
Comment (perf2711) [reply lib/workflows/integrations.js]: added a prefix for each object type

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md:1600]: I've updated the README with more information and examples.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md]: The example works fine.

Not sure what you mean by the describe endpoint? How will it help here? We do not have describes also for new plugins.

**Review by wca — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+
+```
+morgue workflows connection [create | list | get | update | delete] <options>
+morgue workflows integration [create | list | get | update | delete] <options>
+morgue workflows alert [create | list | get | update | delete] <options>
+```
+
+### Managing connections
+
+Managing connections requires an universe to be specified.
```
Comment (rick-bt) [inline README.md:1550]:
```suggestion
Managing connections requires a universe to be specified.
```

File: README.md [hunk 9]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+If it is not set via config or login, specify it by `--universe`.
```
Comment (rick-bt) [inline README.md:1551]:
```suggestion
If it is not set via config or login, specify it with `--universe`.
```

File: README.md [hunk 10]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+
+Example:
+
+```
+morgue workflows connection delete 10
+```
+
+### Managing integrations
+
+Managing integrations requires an universe and a project to be specified.
```
Comment (rick-bt) [inline README.md:1682]:
```suggestion
Managing integrations requires both a universe and a project to be specified.
```

File: README.md [hunk 11]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+If the universe is not set via config or login, specify it by `--universe`.
```
Comment (rick-bt) [inline README.md:1683]:
```suggestion
If universe or project are not set via config or login, specify them with `--universe` and `--project`.
```

File: README.md [hunk 12]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+Specify the project using `--project`.
```
Comment (rick-bt) [inline README.md:1684]:
```suggestion
```

File: README.md [hunk 13]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+    "credentials": {
+      "awsAccessKeyId": "key",
+      "awsSecretAccessKey": "secret"
+    },
+    "attributeList": ["attr1", "attr2"]
+  }
+}
+```
+
+#### Update a integration
```
Comment (rick-bt) [inline README.md:1780]:
```suggestion
#### Update an integration
```

File: README.md [hunk 14]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+
+```
+morgue workflows integration update <options> <integration id>
+```
+
+By using `--from-file`, you can load integration spec from file.
+Arguments provided by CLI will override the file spec, integrationy be not required anymore.
```
Comment (rick-bt) [inline README.md:1787]:
```suggestion
Arguments provided by CLI will override the file spec, but they may not be required anymore.
```

File: README.md [hunk 15]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+
+Example:
+
+```
+morgue workflows integration delete 20
+```
+
+### Managing alerts
+
+Managing alerts requires an universe and a project to be specified.
```
Comment (rick-bt) [inline README.md:1830]:
```suggestion
Managing alerts requires both a universe and a project to be specified.
```

File: README.md [hunk 16]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+If the universe is not set via config or login, specify it by `--universe`.
```
Comment (rick-bt) [inline README.md:1831]:
```suggestion
If universe or project are not set via config or login, specify them with `--universe` and `--project`.
```

File: README.md [hunk 17]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+Specify the project using `--project`.
```
Comment (rick-bt) [inline README.md:1832]:
```suggestion
```

File: README.md [hunk 18]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+By using `--from-file`, you can load alert spec from file.
+Arguments provided by CLI will override the file spec, but they may not be required anymore.
+
+Options:
+
+- `--name` - `string` - alert name, required from CLI or file spec
+- `--condition` - `object` - alert condition, required from CLI or file spec:
+  - `--condition.name` - `string` - can be one of:
+    - `group`,
+    - `trace`,
```
Comment (rick-bt) [inline README.md:1881]: Should we define group and trace?

**Comment by rick-bt** [comment]
Is it not possible to merge the create and manage sections for each topic?

**Comment by rick-bt** [comment]
I tried playing with a table for the options. Looks nicer, I think, but I got stuck on the alert options, which are more complex. But if you can think of a way to format it, this would be more readable.

| CLI option | File spec | Type | Description |
| - | - | - | - |
| `--name` | `watcherName` | string |integration name, required from CLI or file spec
| `--options` | `options` | object | integration options specific to plugin ID, required from CLI or file spec |
| `state` | `state` | `enabled`\|`disabled`\|`stopped` | integration state, optional |

**Review by rick-bt — CHANGES_REQUESTED** [review]
Please consider suggested changes.

File: README.md (hunk 9 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 11 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 12 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 13 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 14 above)
Comment (perf2711) [reply README.md]: updated

File: README.md [hunk 19]
```diff
@@ -1535,6 +1535,476 @@ follows:
...
+Example:
+
+```
+morgue workflows integration update \
+  --project myProject \
+  --options.bucketPath s3://new-bucket-path \
+  --options.attributeList attr1 \
+  20
+```
+
```
[the remark this answers is not in the document]
Comment (perf2711) [reply README.md:1802]: I do not think we need JSON files for update examples, the create examples already cover this.

File: README.md (hunk 16 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 17 above)
Comment (perf2711) [reply README.md]: updated

File: README.md (hunk 18 above)
Comment (perf2711) [reply README.md]: Added some descriptions to them.

**Comment by perf2711** [comment]
>Is it not possible to merge the create and manage sections for each topic?
Not sure what you mean here.

>I tried playing with a table for the options. Looks nicer, I think, but I got stuck on the alert options, which are more complex. But if you can think of a way to format it, this would be more readable.
I updated the README with tables. Let me know what you think.

**Review by rick-bt — APPROVED** [review]

**Pull request merged**